runtime-modules/common/src/costs.rs

```diff
@@ -0,0 +1,135 @@
+    let unusable_free_balance = locks
+        .iter()
+        .map(|lock| {
+            if lock.id == InvitedMemberLockId::get() {
+                Zero::zero()
+            } else {
+                lock.amount
+            }
+        })
+        .max()
+        .unwrap_or_else(Zero::zero);
+    let usable_funds = free_balance.saturating_sub(unusable_free_balance);
+
+    usable_funds >= amount && total_balance >= amount.saturating_add(T::ExistentialDeposit::get())
+}
+
+// Pay a certain fee associated with a transaction (for example: a bloat bond, data size fee)
+// Fees are assumed to be payable w/ invitation-locked balance.
+// Returns the part of amount that was covered with invitation-locked balance (possibly 0).
+pub fn pay_fee<T: frame_system::Config + balances::Config>(
+    payer: &T::AccountId,
+    reciever: Option<&T::AccountId>,
+    amount: T::Balance,
+) -> Result<T::Balance, DispatchError> {
+    if amount == Zero::zero() {
+        return Ok(Zero::zero());
+    }
+
+    let account = T::AccountStore::get(payer);
+
+    ensure!(
+        has_sufficient_balance_for_fees::<T>(payer, amount),
+        DispatchError::Other("pay_fee: Insufficient balance")
+    );
+
+    // Calculate payer's new free balance
+    let new_free_balance = account
+        .free
+        .checked_sub(&amount)
+        .ok_or(DispatchError::Other("pay_fee: Free balance underflow"))?;
+
+    // Set payer's new free_balance
+    let withdraw_imbalance = balances::Pallet::<T>::make_free_balance_be(payer, new_free_balance);
+
+    // Check balance update result
+    // Make sure that the expected amount was withdrawn from payer's free_balance
+    match withdraw_imbalance {
+        SignedImbalance::Negative(negative_imbalance) if negative_imbalance.peek() != amount => {
+            Err(DispatchError::Other(
+                "pay_cost: Unexpected withdrawal imbalance - Amount does not match cost.amount",
+            ))
+        }
+        SignedImbalance::Positive(_) => Err(DispatchError::Other(
+            "pay_cost: Unexpected withdrawal imbalance - Positive",
+        )),
+        _ => Ok(()),
+    }?;
+
+    if let Some(reciever) = reciever {
+        // Deposit funds into reciever's account
+        let deposit_imbalance = balances::Pallet::<T>::deposit_creating(reciever, amount);
+
+        // Check deposit result
+        ensure!(
+            deposit_imbalance.peek() == amount,
+            DispatchError::Other("pay_cost: Unexpected deposit imbalance")
+        );
+    }
+
+    // locked_balance_used = min(account.frozen(Reasons::All).saturating_sub(new_free_balance), amount)
+    let locked_balance_used = account
+        .misc_frozen
+        .max(account.fee_frozen)
+        .saturating_sub(new_free_balance)
+        .min(amount);
+
+    Ok(locked_balance_used)
+}
+
+pub fn burn_from_usable<T: frame_system::Config + balances::Config>(
+    account: &T::AccountId,
+    amount: T::Balance,
+    allow_death: bool,
```

Comment:
> Seems to me that we should just always allow it at the core level, but that the relevant account will always have the sufficient balance prefunded to block it when it should not happen.

There are 2 separate cases when I'm providing `false` as `allow_death`:
1. If the account is never expected to die anyway (that's the `moderate_post` example)
1. If the account could die if we didn't explicitly ensure that the balance is sufficient for it not to die (pay for NFT, buy membership etc.)

In the ideal world it's never actually required for `burn_from_usable` to check this, because we're either assuming the account cannot die (case `1.`) or we already checked the balance to ensure that it won't (case `2.`), so this is just another example of an "extra safety" check.

> curator group paying for channel from user.

When curator group is paying there is no room for allowing/disallowing account death, because the funds are just substracted from the ContentWorkingGroup budget and deposited into previous owner's account. Perhpas you meant member paying for curator group channel?

Reply:
Removed `allow_death` parameter in https://github.com/Joystream/joystream/pull/4164/commits/fe0a624904917b1747c5ec5ae2c63cdfe5e3fc42